File: Leetcode/stack_queue/20_ValidParentheses/Solution.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i<s.length(); ++i){
            Character ch = s.charAt(i);
            if(ch=='(' || ch=='[' || ch=='{'){
                stack.push(ch);
            }else if(!stack.empty() && checkParentheses(stack.peek(), ch)){
                stack.pop();
            }else{
                return false;
            }
        }
        return (stack.empty())? true: false;
    }
        private boolean checkParentheses(Character c1, Character c2){
            return (c1=='(' && c2==')' ) || (c1=='{' && c2=='}') || (c1=='[' && c2==']');
        }
}

File: Leetcode/stack_queue/20_ValidParentheses/Solution.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stk = []
        checkParen = lambda c1, c2: (c1=='(' and c2==')' ) or (c1=='{' and c2=='}') or (c1=='[' and c2==']')
        for char in s:
            if char == '(' or char == '[' or char == '{':
                stk.append(char)
            elif stk and checkParen(stk[-1], char):
                stk.pop()
            else:
                return False
        return True if not stk else False
